ssp fixup with multiple config

juimonen opened the pull request:
Using now SSP BE fixup function to set the dai config. This version is using related kcontrol to choose the config, but can be changed to use other matching method. Tested in up2, works pretty nicely.

juimonen commented:
related fw PR: https://github.com/thesofproject/sof/pull/3926

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -3257,88 +3283,101 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
-	/* configure dai IPC message */
-	hw_config = &cfg->hw_config[i];
+	/* reserve memory for all hw configs, eventually freed by widget */
+	config = kzalloc(number_configs * sizeof(*config), GFP_KERNEL);
```
`kcalloc()`

juimonen replied:
ok, I think I changed already before based on your comments :) so somehow lurking back...

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -3257,88 +3283,101 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
+					config);
 		break;
 	case SOF_DAI_INTEL_DMIC:
 		ret = sof_link_dmic_load(scomp, index, link, cfg, hw_config,
```
this seems to be changing the `hw_config` parameter for all `sof_link_*_load()` functions. You deal with it for SSP but you don't touch others. Before this patch `hw_config` was pointing to the matching HW config:
```
hw_config = &cfg->hw_config[i];
```
now it's the first of them:
```
hw_config = cfg->hw_config;
```
Is this correct? Didn't you mean to change these calls to e.g.
```
ret = sof_link_dmic_load(scomp, index, link, cfg, hw_config + current_config,
```
?

juimonen replied:
I need to parse all hw_configs to be able later to choose 1 of them dynamically. Before none of our topologies included multiple configs. So I guess you could define multiple configs for other dais than SSP -> then we would just discard the others and choose the "default" -> as we did before. So in essence this was previously always sending the first config.

But As I'm doing this for SSP only currently, I'll change the array subscript back for the other dais. If someone wants to add the support for multiple configs for other dais later, the pointer should point to the first config.

juimonen replied:
should match now

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -2876,56 +2884,73 @@ static int sof_link_ssp_load(struct snd_soc_component *scomp, int index,
+	for (current_config = 0; current_config < number_configs; current_config++) {
+		if (cfg->hw_config[current_config].id == cfg->default_hw_config_id)
+			break;
+	}
```
This loop will always find a match because the caller has already done that once... But then you don't actually need to run this search again, just pass `current_config` from the caller?

juimonen replied:
yeah, was just trying to minimize the signature changes... annoying that the dai that I want to actually access is only "dug out" in the ipc sending function. Actually I just want to set the config number related members to dai -> but need to pass down those with quite a route...

juimonen replied:
ok

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -3257,88 +3283,101 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
 		}
 
-		if (i == num_hw_configs) {
+		if (current_config == number_configs) {
```
currently, `current_conf` is only used inside this loop for the ID check. If you change this patch to also pass it to `sof_link_ssp_load()` and use it to calculate the correct `hw_config` for other such calls as I proposed, then it makes sense to have a "distinctive" name for the variable. If however you choose a different solution, just using `i` here should be ok.

juimonen replied:
ok, I'm using it now little bit differently

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -3257,88 +3283,101 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
 
 	/* now load DAI specific data and send IPC - type comes from token */
-	switch (config.type) {
+	switch (config->type) {
```
`config` is an array, so although the above works, you actually meant `config[0].type`, and is it correct? Is it because all types should be the same?

juimonen replied:
yes all types are the same. You actually might be able to set up different types with without error in topology and alsalib tools. haven't tried that. Anyway these are dai config parameters to single dai -> so types should be the same

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -3257,88 +3283,101 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
 	return 0;
+
+err:
+	kfree(config);
```
you have to free `config` in the success case too some time later, or use `devm_kcalloc()`

juimonen replied:
this is freed in L2594 in widget free. config is set to dai_config, it was previously set with kmemdup, so I think I'm not changing the functionality? I'm just reserving the area in an earlier stage than before.

lyakh commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -400,6 +400,101 @@ static const struct snd_soc_dai_ops hda_link_dai_ops = {
+				if (index < number_configs &&
+				    dai->dai_config[index].type == SOF_DAI_INTEL_SSP) {
+					dev_dbg(sdev->dev,
+						"found dai config at index %d\n", index);
```
I think we're trying to use "SSP," "DAI" etc. - in all capitals - in comments and prints - there are several lower case occurrences above.

juimonen replied:
ok. I'll try to unify

juimonen commented:
I'm little bit puzzled why this fails building... this builds for me locally on top of latest topic/sof-dev.

kv2019i commented on sound/soc/sof/topology.c:
```diff
@@ -3257,88 +3283,101 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
 
-		for (i = 0; i < num_hw_configs; i++) {
-			if (cfg->hw_config[i].id == cfg->default_hw_config_id)
+		for (current_config = 0; i < number_configs; current_config++) {
```
Oops, "i" is not initialized here, so this is not right.

juimonen replied:
ok, will fix...

kv2019i commented on sound/soc/sof/topology.c:
```diff
@@ -3257,88 +3283,101 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
+	hw_config = cfg->hw_config;
+	number_configs = le32_to_cpu(cfg->num_hw_configs);
+	if (!number_configs) {
+		if (config->type != SOF_DAI_INTEL_HDA) {
```
config is not initialized here

juimonen replied:
oh yes... this is a good catch, whole thing is for some old wrongly defined topologies... I'll have to do this different way ot preserve the old functionality

kv2019i reviewed (changes requested):
Not a full review, a few error noted during testing.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -400,6 +400,101 @@ static const struct snd_soc_dai_ops hda_link_dai_ops = {
+				index = cdata->chanv[0].value;
+				/* check that the control value is in range and for ssp only */
+				if (index < number_configs &&
+				    dai->dai_config[index].type == SOF_DAI_INTEL_SSP) {
```
what if tomorrow Seppo wants to try multiple configs for microphones? What's so special about the SSP?

juimonen replied:
nothing, I just didn't want to touch other than SSP and keep the original functionality for the other DAI's. IMHO all dai's should support the multiple configs. This is just trying to minimize the damage if someone tries to do some strange configuration for DAI's that don't currently save multiple configs.

ranj063 replied:
why do we need to check, you dont set the BE hw_params for DMICs anyway. So its not like you can even get here for DMICs

juimonen replied:
ok, removing this

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -422,6 +517,7 @@ static struct snd_soc_cdai_ops sof_probe_compr_ops = {
 struct snd_soc_dai_driver skl_dai[] = {
 {
 	.name = "SSP0 Pin",
+	.ops = &ssp_link_dai_ops,
```
are we confusing dai and dailink here?

kv2019i replied:
there's definitely term overload here. the file uses "link" to refer to HDA concept of stream links. This has no direct relation to dailinks of ASoC. For most part this is ok, but the existing function name "hda_link_dai_ops", and ssp_link_dai_ops added here, start to risk a higher degree of confusion. 

With HDA, the hw_params is overriden to pass "HDA link" related info, but the function name could perhaps be something else.

In case of SSP, there's no business to have "link".

Long story short: s/ssp_link_dai_ops/ssp_dai_ops/ and this is cleared.

juimonen replied:
ok

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -400,6 +400,101 @@ static const struct snd_soc_dai_ops hda_link_dai_ops = {
+
+		if (!strcmp(dai->name, sof_dai->cpu_dai_name) &&
+		    substream->stream == sof_dai->comp_dai.direction) {
+			config = ssp_dai_config_control(sdev, sof_dai);
```
ok, I am lost here.

This is just selecting a config without looking at the hw_params.

What happens if the user sets the frame sync to 48kHz and the hw_params is 8Hz rate? 

I just don't get how errors might be handled.

juimonen replied:
If you set your SSP to 48kHz and try to play 8khz to pcm ->  you get "input/output error" etc. so aplay doesn't work as the hw_params don't match the dai. Exactly similar functionality as you would get currently if you try to play without this selection mechanism to 48khz pipeline without SRC. If you have SRC in your pipe things just work.

ranj063 replied:
I thought we're going to match with runtime hw_params and match config that matches no?

juimonen replied:
there will be 3 "part" mechanism: 1. check if there's related control 2. if no control, match to pcm params, 3 if no match, take default. And I'll move the match to fe fixup...

plbossart reviewed (changes requested):
rather confused on the 'matching' which doesn't look at the hw_params, and the hw_params is not modified either so how would it work for a regular codec (outside of the dummy codec mode)?

ranj063 commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -400,6 +400,101 @@ static const struct snd_soc_dai_ops hda_link_dai_ops = {
 	.prepare = hda_link_pcm_prepare,
 };
 
+static struct sof_ipc_dai_config* ssp_dai_config_control(struct snd_sof_dev *sdev,
```
@juimonen can we split this patch into smaller ones please. First add the BE hw_params for SSP DAIs. It wont do anything but just send the default hw_config. In the following patch you can add the logic to match the DAI_CONFIG with the runtime hw_params and send that one. In the last patch (if needed) you can add the logic for kcontrol.

And oh btw, the be_dai_link_fixup() function needs to be picked as well no?

juimonen replied:
ok, chopped the last commit into 2 separate. don't get the comment on be_dai_link_fixup?

juimonen commented:
@lyakh @kv2019i @ranj063 tried to address your comments... hopefully tracked all:
- tried to handle the possible multi hw_config case in other than SSP (we only take 1 even if someone tries to send multiple)
- removed multiple searching for default_config
- i/curr_conf/num_conf for loop confusions fixed
- unified comments for capital DAI/SSP
-  separated be dai fixup config sendind from match function (last commit)

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -3274,71 +3292,88 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
+		/* Currently only SSP supports multiple configs, for other DAI's only default */
+		/* config is parsed and saved */
+		if (common_config.type != SOF_DAI_INTEL_SSP)
+			num_conf = 1;
```
isn't it an error to specify multiple configs for non-SSP DAIs? Maybe check and return `-EINVAL` before looking for a current configuration?

juimonen replied:
hmm... I was thinkin how it "works" currently that if for some reason multiple configs are defined, we only take 1 which is pointed out by default config

juimonen replied:
ok, got some differentiating comments from @lyakh and @ranj063. I'm know changing it so that anyway memory for all configs is reserved and parsed, but in case of other than SSP, only default (i.e. 1 piece) will be used.

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -2868,64 +2869,81 @@ static int sof_set_dai_config(struct snd_sof_dev *sdev, u32 size,
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
 	struct snd_soc_tplg_private *private = &cfg->priv;
+	int num_conf = le32_to_cpu(cfg->num_hw_configs);
```
this can be 0, which you fix in `sof_link_load()` to 1, but here you don't. Looks like you have to pass `num_config` too?

juimonen replied:
this is already SSP type and the "link fix" you mentioned should return error if "SSP & num_conf==0"

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -3274,71 +3292,88 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
-	/* configure dai IPC message */
-	hw_config = &cfg->hw_config[i];
+	/* Reserve memory for all hw configs, eventually freed by widget */
+	config = kcalloc(num_conf, sizeof(*config), GFP_KERNEL);
```
You explained why this isn't a memory leak, but I think there's still a possibility here: what if `if (!found)` is true in `sof_set_dai_config_multi()`? It would still return 0, and `config` wouldn't be assigned to anything?

juimonen replied:
hmm ok, good point, have to say I'm not sure in which kind of conditions this would be happen, but better be sure here. will change.

ranj063 replied:
this is a valid point. You do need to free the config if no DAI is found in tpology matching the DAI_LINK.

juimonen replied:
ok

lyakh commented on sound/soc/sof/pcm.c:
```diff
@@ -673,10 +673,10 @@ int sof_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd, struct snd_pcm_hw_pa
+		rate->min = dai->dai_config[dai->current_config].ssp.fsync_rate;
+		rate->max = dai->dai_config[dai->current_config].ssp.fsync_rate;
+		channels->min = dai->dai_config[dai->current_config].ssp.tdm_slots;
+		channels->max = dai->dai_config[dai->current_config].ssp.tdm_slots;
```
shouldn't this hunk be in the previous commit?

juimonen replied:
yes maybe... this is a new addition and you might be right.

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -3274,71 +3292,88 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
+		config[0].format = hw_config[curr_conf].fmt;
+		config[0].type = common_config.type;
+		config[0].dai_index = common_config.dai_index;
+		hw_config += curr_conf;
```
Is this else part really necessary? If num_conf is 1, curr_conf will be 0 no?

kv2019i replied:
Ack, I think you can just do the for loop.

juimonen replied:
not necessary if you have defined multiple configs in your topology and you default conf doesn't point to the first one. This is unlikely but possible. I set num_conf forcibly to 1 if this is not SSP dai. That was the previous functionality : discard other possible defined configs than the default conf.

ranj063 replied:
oh that is confusing. so you're saying that there're multiple configs but you force it down to 1. Lets not do that. I think the better way is to just leave the multiple configs and set the default config.

juimonen replied:
ok

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2868,64 +2869,81 @@ static int sof_set_dai_config(struct snd_sof_dev *sdev, u32 size,
+	ret = sof_parse_token_sets(scomp, &config->ssp, ssp_tokens,
+				   ARRAY_SIZE(ssp_tokens), private->array,
+				   le32_to_cpu(private->size),
+				   num_conf, size);
```
dont you have to copy this common data into all config's?

juimonen replied:
this is not common and I'm using parse_token_sets(num_conf) -> all confs are copied to correct places

ranj063 replied:
Sorry @juimonen, I dont get this. You only have one set of ssp_tokens in topology , right? So does parse_token_sets() copy the same set to multiple configs?

lyakh replied:
that happens when a pointer is used instead of an array element, yes ;-) If it had been written appropriately as
```
ret = sof_parse_token_sets(scomp, config[0].ssp, ssp_tokens,
			   ARRAY_SIZE(ssp_tokens), private->array,
			   le32_to_cpu(private->size),
			   num_conf, size);
```
(note `config[0].` instead of `config->`) then it would have been obvious, that other array members have to be used too...

juimonen replied:
@lyakh I have to say at least for me it would not make it any more clear, but that could be just me. parse_token_sets is quite well documented (I think you asked me to do that way back) and num_conf is given as argument (how many items to parse). The pointer to the beginning of "copy area" could be given in many ways I guess... anyway I can change this if you think it is useful.

ranj063 commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -400,6 +400,50 @@ static const struct snd_soc_dai_ops hda_link_dai_ops = {
+	config = &dai->dai_config[current_config];
+
+	/* return the only possible config if only 1 config present */
+	if (number_configs < 2)
```
why not be explicit and say num_configs == 1

juimonen replied:
ok, yes.

kv2019i commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -419,11 +440,19 @@ static int ssp_dai_hw_params(struct snd_pcm_substream *substream,
+				config = &sof_dai->dai_config[sof_dai->current_config];
+			/* try to match to pcm params */
+			if (!config)
+				config = ssp_dai_config_pcm_params_match(sdev, sof_dai, params);
```
This is interesting. @plbossart @ranj063 @lyakh  We discussed with Jaska and basicly this matching part can be done here (in DAI/BE hw_params), but it could also be done in sof/pcm.c:sof_pcm_dai_link_fixup().

It would seem from ASoC point of view both would be good.

Order-wise fixup is called first, so it would actually be cleanest solution to have the matching logic there, in sof/pcm.c. 

But if fixup is meant primarily to adjust for information coming from topology, then the matching part does not belong there.

Any insights on which is the better place?

ranj063 replied:
Yes, should be done in sof_pcm_dai_link_fixup()

juimonen replied:
ok, moving there

kv2019i reviewed (changes requested):
Some comments, see inline. I'd prefer if you split this into two PRs. One with only runtime matching, and a separate PR that has the ASoC change and kcontrol. Otherwise we might again have too much to review at once.

ranj063 commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -419,11 +440,19 @@ static int ssp_dai_hw_params(struct snd_pcm_substream *substream,
+				dev_err(sdev->dev, "error: DAI config index out of range\n");
 				return -EINVAL;
 			}
+			/* we can return immediately as there are no choises */
```
choises -> choices

juimonen replied:
ok

ranj063 commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -419,11 +440,19 @@ static int ssp_dai_hw_params(struct snd_pcm_substream *substream,
+				config = ssp_dai_config_pcm_params_match(sdev, sof_dai, params);
+			/* last resort set default config */
+			if (!config)
+				config = &sof_dai->dai_config[sof_dai->current_config];
```
why not flip the logic so you do the params matching first. If no match found, set default config. You do the same thing is the first and the third if's anyway

juimonen replied:
this will now change

ranj063 commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -447,6 +480,9 @@ static int ssp_dai_hw_params(struct snd_pcm_substream *substream,
 				config = &sof_dai->dai_config[sof_dai->current_config];
+			/* check for possible enum control set by user */
+			if (!config)
+				config = ssp_dai_config_control_match(sdev, sof_dai);
```
this seems moot now no? You'd ignore this and go ahead and match with pcm params anyway? So whats the point of the kcontrol then?

juimonen replied:
ok, this part will change

juimonen commented:
@lyakh @ranj063 @kv2019i @plbossart updated. Apologies for the update-storm, but this keeps getting refined...
changes in this version:
- removed kcontrol match and related patches, now has only 3 patches (multi_ssp parsing, ssp be fixup for config send, fe pcm params match)
- tried to address all comments
- simplified: this PR parses multiple SSP configs if found, in stream start it matches those with pcm params and sends the config in be fixup

kv2019i commented:
Removed "[RFC]" from title, this is for real now, folks. :)

kv2019i commented on sound/soc/sof/topology.c:
```diff
@@ -2861,71 +2869,89 @@ static int sof_set_dai_config(struct snd_sof_dev *sdev, u32 size,
 	 * or none of them. Here print a warning message to notify user
 	 */
 	if (!found) {
+		kfree(config);
```
I guess this is safe, but it's a bit non-obvious to see this is correct as the alloc has been done two functions up the callstack. Would it be possible to return error here, so the memory would be freed in sof_link_load() ? That would be easier to veriy-to-be-correct by reading the code.

ranj063 replied:
I agree, I think we should consolidate the return paths for all errors. Why deal with this one separately?

juimonen replied:
ok, but what is the functionality of this whole thing? This code was there previously and it is not returning error in this case... so is it ok to return error here? the parsed config is "thrown away" with this and previous versions so what is the point here?

ranj063 replied:
yeah it didnt make sense before. So lets fix it

juimonen replied:
I changed this now to return an error.

kv2019i commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -400,6 +400,49 @@ static const struct snd_soc_dai_ops hda_link_dai_ops = {
 	.prepare = hda_link_pcm_prepare,
```
The lower case "dai" pops up a bit in commit message. If you look at git log history for ASoC, it's all mixed and matches (and "hw" is an acronym as well), so maybe this is ok. People will understand what this means.

juimonen replied:
ok changing

kv2019i reviewed (changes requested):
Excellent. thanks @juimonen ! This is now very clean and easy to read. I had one comment and the CI build uncovered this:

```
sound/soc/sof/intel/hda-dai.c:468:10: error: â€˜ssp_dai_opsâ€™ undeclared here (not in a function); did you mean â€˜sof_icl_opsâ€™?
  .ops = &ssp_dai_ops,
          ^~~~~~~~~~~
```

Otherwise, this looks good to go!

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2811,12 +2811,14 @@ static void sof_dai_set_format(struct snd_soc_tplg_hw_config *hw_config,
+				    int num_conf, int curr_conf)
 {
 	struct snd_sof_dai *dai;
 	int found = 0;
+	int i = 0;
```
no need to init ?

juimonen replied:
ok, will remove

ranj063 commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -414,6 +414,49 @@ static struct snd_soc_cdai_ops sof_probe_compr_ops = {
 #endif
 #endif
 
+static int ssp_dai_hw_params(struct snd_pcm_substream *substream,
```
comment on teh commit message. This isnt adding the BE fixup function but rather the BE hw_params callback right?

ranj063 replied:
and we should also mention in the commit message that this patch would introduce a new IPC that doesnt exist today.

juimonen replied:
yes will fix

ranj063 commented on sound/soc/sof/pcm.c:
```diff
@@ -619,6 +619,33 @@ static int sof_pcm_new(struct snd_soc_component *component,
+
+	/* we might have several dai's associated with same link, need to loop all */
+	list_for_each_entry(dai, &sdev->dai_list, list) {
+		if (dai->name && (strcmp(name, dai->name) == 0)) {
```
!strcmp(...)?

juimonen replied:
ok, this was a direct copy from existing function but can change.

ranj063 commented on sound/soc/sof/pcm.c:
```diff
@@ -673,6 +701,9 @@ int sof_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd, struct snd_pcm_hw_pa
 	switch (dai->dai_config->type) {
 	case SOF_DAI_INTEL_SSP:
+		/* search for config to pcm params match, if not found use default */
+		ssp_dai_config_pcm_params_match(sdev, (char *)rtd->dai_link->name, params);
```
you dont check the return value here, so why bother with the return -EINVAL at all in the function?

juimonen replied:
This is for the future, if I need to do comparison with several different match rules/functions, now I'm just ignoring the error -> go with default config

ranj063 commented on sound/soc/sof/pcm.c:
```diff
@@ -619,6 +619,33 @@ static int sof_pcm_new(struct snd_soc_component *component,
+		}
+	}
+
+	return ret;
```
You dont do anything with the return value? nor do you modify current_config in case no match is found? It is left as default and so what is the outcome when no match is found?

juimonen replied:
it is using default

juimonen replied:
@ranj063 so the idea is that in the future I can call this function: if returns NULL I can search for something else, and so on.

juimonen commented:
@ranj063 parse_token_sets copies different items to multiple consecutive configs. Because it knows the size, as all tokens have been parsed it "jumps" to the beginning of next config in the array.

juimonen commented:
@ranj063 tried to incorporate your suggestion. Match function is now returning void and I took one level of indentation away to make it a little bit cleaner.

ranj063 commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -414,6 +414,49 @@ static struct snd_soc_cdai_ops sof_probe_compr_ops = {
 #endif
 #endif
 
+static int ssp_dai_hw_params(struct snd_pcm_substream *substream,
```
prefix needs to be fixed to ASoC: SOF: Intel: HDA

juimonen replied:
ok

ranj063 commented on sound/soc/sof/pcm.c:
```diff
@@ -619,6 +619,27 @@ static int sof_pcm_new(struct snd_soc_component *component,
+		for (i = 0; i < dai->number_configs; i++) {
+			config = &dai->dai_config[i];
+			if (config->ssp.fsync_rate == params_rate(params)) {
+				dev_dbg(sdev->dev, "DAI config pcm params match at index %d\n", i);
```
not sure this is a very helpful log. If anything index would confuse me with DAI index.

juimonen replied:
ok, can try to make this more readable, anyway the dbg msg here is _really_ useful

ranj063 commented on sound/soc/sof/pcm.c:
```diff
@@ -619,6 +619,27 @@ static int sof_pcm_new(struct snd_soc_component *component,
+	struct snd_sof_dai *dai;
+	int i;
+
+	/* we might have several dai's associated with same link, need to loop all */
```
this part is not very clear too, why all DAI's, shouldnt we check for direction too?

juimonen replied:
So I understood this is for all dai's in a link. So actually the calling function call sof_find_dai which finds only SSP_IN.
If you can come up how I can dig up the dir from both dai and pcm hw params, I can do that

ranj063 replied:
hmm ok for SSP, I guess the config is the same for both playback and cpature DAI. So instead of saying several, can we change to say both playback/capture DAIs?

juimonen replied:
ok. I tried to make the comment better and I renamed the parameter to link_name to make it more clear what is happening here...

lyakh commented on sound/soc/sof/pcm.c:
```diff
@@ -619,6 +619,30 @@ static int sof_pcm_new(struct snd_soc_component *component,
+			config = &dai->dai_config[i];
+			if (config->ssp.fsync_rate == params_rate(params)) {
+				dev_dbg(sdev->dev, "DAI config %d matches pcm hw params\n", i);
+				dai->current_config = i;
```
you don't want to break the inner loop here?

juimonen replied:
no, I have to go through all DAI's associated with the same link.

ujfalusi replied:
So, you want to find the last config which matches?
The loop can be written this way to achieve the same result:
```
for (i = dai->number_configs; i; i--) {
	config = &dai->dai_config[i - 1];
	if (config->ssp.fsync_rate == params_rate(params)) {
		dev_dbg(sdev->dev, "DAI config %d matches pcm hw params\n", i);
		dai->current_config = i - 1;
		break;
	}
}
```
It is true that you would not have dev_dbg() prints for all matches, but if this is the intention I would update the comment to reflect it.

ranj063 replied:
no I think the idea is to find the first matching config but set the config for all matching DAIs. So the loop looks good to me

lyakh reviewed (changes requested):
please have a look at @ranj063's comment too

kv2019i reviewed (approved):
Thanks @juimonen and reviewers, this looks good now!

kv2019i commented on sound/soc/sof/topology.c:
```diff
@@ -2868,64 +2876,81 @@ static int sof_set_dai_config(struct snd_sof_dev *sdev, u32 size,
+		config[i].ssp.fsync_rate = le32_to_cpu(hw_config[i].fsync_rate);
+		config[i].ssp.tdm_slots = le32_to_cpu(hw_config[i].tdm_slots);
+		config[i].ssp.tdm_slot_width = le32_to_cpu(hw_config[i].tdm_slot_width);
+		config[i].ssp.mclk_direction = hw_config[i].mclk_direction;
```
@juimonen  I hit en error with 16khz test and error would seem to be that config[i].ssp.sample_valid_bits is zero. Should be copied here as well?

kv2019i replied:
@juimonen false alarm, topology had a oldstyle declaration for this SSP and that caused the error -> still good

kv2019i reviewed (approved):
Tested with latest 1.7-rc and a TGL product topology that uses multiple dai configs for a SSP.

plbossart reviewed (commented):
I have mixed feelings here.
One one hand the code looks fine.

But I also see that the changes anchor the confusion in SOF between DAI and DAI_LINK. The hw_configs and defaults are properties of a DAI_LINK: see https://elixir.bootlin.com/linux/latest/source/include/uapi/sound/asoc.h#L545.

But we keep referring to DAIs and in the last patch we introduce hw_params for the DAIs.

That makes me think we should clean this up and stop conflating the two terms.

juimonen commented:
> I have mixed feelings here.
> One one hand the code looks fine.
> 
> But I also see that the changes anchor the confusion in SOF between DAI and DAI_LINK. The hw_configs and defaults are properties of a DAI_LINK: see https://elixir.bootlin.com/linux/latest/source/include/uapi/sound/asoc.h#L545.
> 
> But we keep referring to DAIs and in the last patch we introduce hw_params for the DAIs.
> 
> That makes me think we should clean this up and stop conflating the two terms.

@plbossart I'm also confused by the terminology. I can go through the commit messages and comments in this PR to try to unify the terms. However I don't think code clean up is part of this, I've been many times told to remove all non-related parts, so I should not mix those with functionality.

That said, it is clear to me when doing this PR, that the link loading part differs between DAI types and we have quite a lot of strange legacy code, which complicates the code and no one has an idea why they are there originally and do we need those today.

lyakh commented:
@juimonen interesting, why did you decide to return to `kmemdup()`?

juimonen commented:
> @juimonen interesting, why did you decide to return to `kmemdup()`?

@lyakh sorry not to mentioned that... yesterday the CI showed capture error's in many places. So actually the config was set as common item to all dai's in a link. Like with SSP it was set to SSP_IN and SSP_OUT. When set parameters like current_config for SSP_IN -> set also for SSP_OUT. This was not a problem with audio params, mainly with some link index. Anyway every DAI needs to have their own copy of the configs. kudos for our CI

juimonen commented:
@lyakh so in a way the memory allocs in parsing is not very "neat", but there's a lot of legacy code I need to take into account.
1) static common_config struct parsed first, as we need to get the DAI type for some legacy num_conf mangling
2) then I can reserve memory for possible multiple configs -> and parse it
3) finally I need to copy this parsed config with kmemdup to all related DAI's in a link

kv2019i commented:
@plbossart wrote:
> But I also see that the changes anchor the confusion in SOF between DAI and DAI_LINK. The hw_configs and defaults are properties of a DAI_LINK: see https://elixir.bootlin.com/linux/latest/source/include/uapi/sound/asoc.h#L545.
[...]
> But we keep referring to DAIs and in the last patch we introduce hw_params for the DAIs.
[...]
> That makes me think we should clean this up and stop conflating the two terms.

I'm not sure what that clean up would be. hw_params is a valid method on DAIs. The hda-dai.c hwparams call seems ok from this perspective.

In sof/pcm.c, we use snd_soc_component_driver be_fixup, but that seems established use and is not mixing DAI and DAILINKS (at least not more).

So what are we anchoring? I probably missed the point :)

plbossart commented:
> In sof/pcm.c, we use snd_soc_component_driver be_fixup, but that seems established use and is not mixing DAI and DAILINKS (at least not more).
> 
> So what are we anchoring? I probably missed the point :)

We are absolutely mixing dai and dai_link in the fixup code:

````
struct snd_sof_dai *snd_sof_find_dai(struct snd_soc_component *scomp,
				     const char *name)
{
	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
	struct snd_sof_dai *dai;

	list_for_each_entry(dai, &sdev->dai_list, list) {
		if (dai->name && (strcmp(name, dai->name) == 0))
			return dai;
	}

	return NULL;
}

int sof_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd, struct snd_pcm_hw_params *params)
{
	struct snd_interval *rate = hw_param_interval(params,
			SNDRV_PCM_HW_PARAM_RATE);
	struct snd_interval *channels = hw_param_interval(params,
						SNDRV_PCM_HW_PARAM_CHANNELS);
	struct snd_mask *fmt = hw_param_mask(params, SNDRV_PCM_HW_PARAM_FORMAT);
	struct snd_soc_component *component =
		snd_soc_rtdcom_lookup(rtd, SOF_AUDIO_PCM_DRV_NAME);
	struct snd_sof_dai *dai =
		snd_sof_find_dai(component, (char *)rtd->dai_link->name);
````

I guess it works because dai links and dais happen to have the same name? How reliable is this?

And now we are making it worse by adding hw_params on dais to deal with changes in configurations of the dailink.

When we mix-up concepts and layers, sooner or later things will break. Let's use the proper abstractions and interfaces.

juimonen commented:
@plbossart hmm.. but I mean there's is no dai_link concept in the dsp? I understand dai_link means the actual hw and we might have several dai (input/output) associated with the same dai_link. Only way to configure the dai_link is to send ipc to dsp to these possibly multiple dai's. So this translation needs to happen somewhere? If you are doing something to the dai_link -> you are effectively configuring the dai component in the dsp.

juimonen commented:
@plbossart don't get me wrong, to me this is also confusing. I just don't know do we have at this point somewhere mapping existing that this dai belongs to this dai link... other than having the same name.

ranj063 commented:
> I guess it works because dai links and dais happen to have the same name? How reliable is this?
> 
@plbossart it doesnt happen to work. The driver is responsible for making sure it works. 
https://github.com/thesofproject/linux/blob/d289bfd938cccef124748e0e9d89ca6a95fcd016/sound/soc/sof/topology.c#L1386
Of course, topology could be broken in which case, this will fail and we will catch the error.

> And now we are making it worse by adding hw_params on dais to deal with changes in configurations of the dailink.
> 
> When we mix-up concepts and layers, sooner or later things will break. Let's use the proper abstractions and interfaces.

But the DAI and the DAI_link are intertwined by nature isnt it? A DAI_link is meaningless is there isnt a DAI associated with it. And converserly, a DAI without a DAI_link wouldnt work either. But I agree that maybe the sort of searching and matching we do is hairy and maybe we can simplify this.

plbossart commented:
> > I guess it works because dai links and dais happen to have the same name? How reliable is this?
> 
> @plbossart it doesnt happen to work. The driver is responsible for making sure it works.
> https://github.com/thesofproject/linux/blob/d289bfd938cccef124748e0e9d89ca6a95fcd016/sound/soc/sof/topology.c#L1386
> 
> 
> Of course, topology could be broken in which case, this will fail and we will catch the error.
> > And now we are making it worse by adding hw_params on dais to deal with changes in configurations of the dailink.
> > When we mix-up concepts and layers, sooner or later things will break. Let's use the proper abstractions and interfaces.
> 
> But the DAI and the DAI_link are intertwined by nature isnt it? A DAI_link is meaningless is there isnt a DAI associated with it. And converserly, a DAI without a DAI_link wouldnt work either. But I agree that maybe the sort of searching and matching we do is hairy and maybe we can simplify this.

it's a different level, a dailink may have more than one dai. We already use this for SoundWire, so I am really nervous about the 1:1 thinking.

Note that this also true in topology files, DAI_CONFIG is really DAI_LINK_CONFIG.

plbossart commented:
> @plbossart hmm.. but I mean there's is no dai_link concept in the dsp? I understand dai_link means the actual hw and we might have several dai (input/output) associated with the same dai_link. Only way to configure the dai_link is to send ipc to dsp to these possibly multiple dai's. So this translation needs to happen somewhere? If you are doing something to the dai_link -> you are effectively configuring the dai component in the dsp.

The dai-link connects two dais. So when you change the configuration, it has to be on both dais, the hw_params need to be understood by both sides, the cpu-dai and the codec-dai. You do not send an IPC to the DSP in isolation, it's part of a broader reconfiguration that involves the codec part as well. You don't see it for BT because the codec is dummy but nothing in this PR is restricted to dummy codecs.

plbossart commented:
> > > I guess it works because dai links and dais happen to have the same name? How reliable is this?
> > 
> > 
> > @plbossart it doesnt happen to work. The driver is responsible for making sure it works.
> > https://github.com/thesofproject/linux/blob/d289bfd938cccef124748e0e9d89ca6a95fcd016/sound/soc/sof/topology.c#L1386
> > 
> > Of course, topology could be broken in which case, this will fail and we will catch the error.
> > > And now we are making it worse by adding hw_params on dais to deal with changes in configurations of the dailink.
> > > When we mix-up concepts and layers, sooner or later things will break. Let's use the proper abstractions and interfaces.
> > 
> > 
> > But the DAI and the DAI_link are intertwined by nature isnt it? A DAI_link is meaningless is there isnt a DAI associated with it. And converserly, a DAI without a DAI_link wouldnt work either. But I agree that maybe the sort of searching and matching we do is hairy and maybe we can simplify this.
> 
> it's a different level, a dailink may have more than one dai. We already use this for SoundWire, so I am really nervous about the 1:1 thinking.
> 
> Note that this also true in topology files, DAI_CONFIG is really DAI_LINK_CONFIG.

I am also very nervous about using the name as a pointer.

https://github.com/thesofproject/linux/blob/d289bfd938cccef124748e0e9d89ca6a95fcd016/sound/soc/sof/topology.c#L1426

This probably works for FEs since there's only one dai for the FE, but on the BE something doesn't make sense.

kv2019i commented:
@plbossart wrote:
> it's a different level, a dailink may have more than one dai. We already use this for SoundWire, so I am really nervous 
> about the 1:1 thinking.

So I was wondering how this can even work today. But it would seem sof_link_alh_load() is not using snd_sof_find_dai(). For SoundWire, config is set to all matching DAIs:

```
       /* set config for all DAI's with name matching the link name */
       ret = sof_set_dai_config(sdev, size, link, config);
```

... so in practise, the find_dai() lookup is only used for link types where we don't support anything else but 1:1 topologies.

> Note that this also true in topology files, DAI_CONFIG is really DAI_LINK_CONFIG.

Hmm, but these properies are not maintained by the link. The dai-link maintains the parameter ranges (snd_soc_pcm_stream), number of codec/CPU DAIs, playback/capture state, symmetry, etc... Specifying which clk source to use, seems to be a property of the CPU DAI. Or it's Friday, and time for me to close the laptop now.

plbossart commented:
> > Note that this also true in topology files, DAI_CONFIG is really DAI_LINK_CONFIG.
> 
> Hmm, but these properies are not maintained by the link. The dai-link maintains the parameter ranges (snd_soc_pcm_stream), number of codec/CPU DAIs, playback/capture state, symmetry, etc... Specifying which clk source to use, seems to be a property of the CPU DAI. Or it's Friday, and time for me to close the laptop now.

Nope, e.g. for SSP the blck, etc are standard dai-link topology config definitions, which we extended with SOF-specific ones.

link configuration:
https://elixir.bootlin.com/linux/latest/source/include/uapi/sound/asoc.h#L338

used here:
https://elixir.bootlin.com/linux/latest/source/include/uapi/sound/asoc.h#L545

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -414,6 +414,48 @@ static struct snd_soc_cdai_ops sof_probe_compr_ops = {
+		if (!strcmp(dai->name, sof_dai->cpu_dai_name) &&
+		    substream->stream == sof_dai->comp_dai.direction) {
+			config = &sof_dai->dai_config[sof_dai->current_config];
+			if (!config) {
```
This check does not make much sense.
I think you want to check if sof_dai->dai_config is not NULL, I hope that sof_dai->current_config is granted to be sanely between 0 and number_configs?

juimonen replied:
hmm, I'm still puzzled by some legacy processing we do in topology.c where some number_config is 0 but there still appears to be a config... anyway your suggestion is good, will change. And anyway I'm setting the current_config so it should be good.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2832,21 +2834,27 @@ static int sof_set_dai_config(struct snd_sof_dev *sdev, u32 size,
+
+			dai->number_configs = num_conf;
+			dai->current_config = curr_conf;
+			dai->dai_config = kmemdup(config, size * num_conf, GFP_KERNEL);
```
You should still check for OoM with if (!dai->dai_config) return -ENOMEM;

juimonen replied:
yes, ok

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -3274,71 +3299,75 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
 	default:
 		dev_err(scomp->dev, "error: invalid DAI type %d\n",
-			config.type);
+			config->type);
```
It would be better to use common_config.type instead config->type.
The switch also uses common_config.type...

juimonen replied:
ok, will change

kv2019i commented:
@plbossart wrote:
> @kv2019i wrote:
>> Hmm, but these properies are not maintained by the link. The dai-link maintains the parameter ranges 
>> (snd_soc_pcm_stream), number of codec/CPU DAIs, playback/capture state, symmetry, etc... Specifying which clk source to >> use, seems to be a property of the CPU DAI. 
>
> Nope, e.g. for SSP the blck, etc are standard dai-link topology config definitions, which we extended with SOF-specific ones.
> link configuration:
> https://elixir.bootlin.com/linux/latest/source/include/uapi/sound/asoc.h#L338

Hmm, with "clk-source" I was thinking of something more DAI specific, but you are right, the "snd_soc_tplg_hw_config" is defined for a link. The reverse also applies, all these link properties need to be programmed to the DAIs of both sides of the links. 

So what do we need to change? The mixups seem to be all over the place. E.g. our "sof_dai_tokens" has SOF_TKN_DAI_INDEX but in documentation it is:
```
dnl DAI_CONFIG(type, idx, link_id, name, sai_config/esai_config/ssp_config/dmic_config)
....
`»      »       SOF_TKN_DAI_INDEX'»     »       STR($2)
```

These should be all be renamed to DAI_LINK_CONFIG and DAI_LINK_INDEX to be consistent right.

But if you make the mental exercise and change these:
  - sdev->dai_list -> sdev->dai_link_list
  -  snd_sof_find_dai() -> snd_sof_find_dai_link()

... the current patch is really not changed. In the sof_pcm_dai_link_fixup(), we are only dealing with dai-links and properties of the link. Once the fixup is done, ASoC core will call hw_params on all entities, including the CPU DAI (via the new hw_params). 

This seems valid. Is this the path to take?

kv2019i commented:
@kv2019i wrote:
> But if you make the mental exercise and change these:
> 
>     * sdev->dai_list -> sdev->dai_link_list
>     * snd_sof_find_dai() -> snd_sof_find_dai_link()

And here's the non-mental version as RFC:
https://github.com/thesofproject/linux/pull/2803

kv2019i commented:
@plbossart and @lyakh: I propose we go ahead and merge this. You can see my long comment on "struct snd_sof_dai" in https://github.com/thesofproject/linux/pull/2803#issuecomment-804878188 ... coming back to this patch, I don't think this patch is adding any new assumptions to the mix. 

Specifically: 
 1) hda-dai.c hw_params - for a given "snd_soc_dai *dai", we have a SOF specific method to access the alternative sets of dai-link properties
       -> SSP CPU DAI hw_params seems to be the correct place to send the DAI_CONFIG IPC

2) topology.c link_ssp_load changes
       -> we need to parse the SOF additional tplg fields... if we decide to store dai-link config to another struct, this won't really change this patch

We could of course have other reasons to block the patch, but the confusion around "struct snd_sof_dai" would seem orthogonal to this patch.

ranj063 reviewed (approved):
looks good to go @juimonen. Thanks!

plbossart reviewed (approved):
this looks good enough to me for now. I still want to continue the discussion on dai/dailinks, and it'd be nice to see if we can modify the HDMI topologies to finally allow for 192kHz multichannel instead of the fixed 48kHz 2 channels w/ HDaudio.

kv2019i commented:
@lyakh can you do one more round?
